server/src/main/java/io/druid/server/QueryResourceV3.java

```diff
@@ -0,0 +1,159 @@
+ *  under the License.
+ * /
+ *
+ */
+
+package io.druid.server;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.CountingOutputStream;
+import com.google.inject.Inject;
+import com.metamx.common.guava.Yielder;
+import com.metamx.emitter.service.ServiceEmitter;
+import io.druid.guice.annotations.Json;
+import io.druid.guice.annotations.Smile;
+import io.druid.query.DruidMetrics;
+import io.druid.query.Query;
+import io.druid.query.QuerySegmentWalker;
+import io.druid.query.QueryToolChest;
+import io.druid.query.QueryToolChestWarehouse;
+import io.druid.server.initialization.ServerConfig;
+import io.druid.server.log.RequestLogger;
+import io.druid.server.security.AuthConfig;
+import org.joda.time.DateTime;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.Path;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.StreamingOutput;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Map;
+
+/**
+ */
+@Path("/druid/v3/")
+public class QueryResourceV3 extends QueryResource
+{
+  private final ServiceEmitter emitter;
+  private final ObjectMapper jsonMapper;
+  private final RequestLogger requestLogger;
+
+  @Inject
+  public QueryResourceV3(
+      QueryToolChestWarehouse warehouse,
+      ServerConfig config,
+      @Json ObjectMapper jsonMapper,
+      @Smile ObjectMapper smileMapper,
+      QuerySegmentWalker texasRanger,
+      ServiceEmitter emitter,
+      RequestLogger requestLogger,
+      QueryManager queryManager,
+      AuthConfig authConfig
+  )
+  {
+    super(warehouse, config, jsonMapper, smileMapper, texasRanger, emitter, requestLogger, queryManager, authConfig);
+    this.emitter = emitter;
+    this.jsonMapper = jsonMapper;
+    this.requestLogger = requestLogger;
+  }
+
+  @Override
+  protected Response buildQueryResponse(
+      final HttpServletRequest req,
+      final Query query,
+      final QueryToolChest toolChest,
+      final ObjectWriter jsonWriter,
+      final String contentType,
+      final long startTime,
+      final Yielder yielder,
+      final Map<String, Object> responseContext
+  )
+      throws IOException
+  {
+    try {
+      return Response
+          .ok(
+              new StreamingOutput()
+              {
+                @Override
+                public void write(OutputStream outputStream) throws IOException, WebApplicationException
+                {
+                  CountingOutputStream os = new CountingOutputStream(outputStream);
+
+                  //jsonSerializer would close the yielder.
+                  jsonWriter.writeValue(
+                      os,
+                      ImmutableMap.of(
+                          "result", yielder,
+                          "context", responseContext
```

Comment:
we might want to put something in the context that can only be obtained after iterating through the result sequence.... in that case, it would be helpful to have context after the results are streamed (so that context can be computed while results are streamed)